nsdb-core/src/main/scala/io/radicalbit/nsdb/index/AbstractStructuredIndex.scala

```diff
@@ -204,4 +205,58 @@ abstract class AbstractStructuredIndex extends Index[Bit] with TypeSupport {
     */
   def getMinGroupBy(query: Query, schema: Schema, groupTagName: String): Seq[Bit] =
     headItemOfGroup(query, schema, groupTagName, last = false, _valueField)
+
+  /**
+    * Group query result by groupTagName and return the unique values for each group.
+    * @param query query to be executed before grouping.
+    * @param schema bit schema.
+    * @param groupTagName tag used to group by.
+    */
+  def uniqueValues(query: Query, schema: Schema, groupTagName: String): Seq[Bit] = {
+    val groupSelector = schema.fieldsMap(groupTagName).indexType match {
+      case VARCHAR() => new TermGroupSelector(groupTagName)
+      case _         => new TermGroupSelector(s"$groupTagName$stringAuxiliaryFieldSuffix")
+    }
+
+    val searcher = this.getSearcher
+
+    val firstPassGroupingCollector =
+      new FirstPassGroupingCollector[BytesRef](groupSelector, new Sort(), maxGroups)
+
+    searcher.search(query, firstPassGroupingCollector)
+
+    val groupsOpt = Option(firstPassGroupingCollector.getTopGroups(0))
+
+    groupsOpt match {
+      case Some(inputGroups) =>
+        val distinctValuesCollector =
+          new DistinctValuesCollector[BytesRef, BytesRef](
+            firstPassGroupingCollector.getGroupSelector,
+            inputGroups,
+            new TermGroupSelector(s"${_valueField}$stringAuxiliaryFieldSuffix"))
+        searcher.search(query, distinctValuesCollector)
+
+        val groups = distinctValuesCollector.getGroups.asScala
```

Comment:
are we able to work directly onto the original collection? I believe `asScala` will require O(n) pass-through.

Reply:
I'm not sure because that variable is mapped afterwards into a sequence of bits that are returned by the method. At some point, we need to convert anyway. What's your thought? Unfortunately Lucene is a java library, therefore using it will lead us to manage java scala interoperability.